courses/codepeer/010_codepeer.rst

```diff
@@ -0,0 +1,2666 @@
+   :number-lines: 1
+
+   function Dup (X : Integer) return Integer is
+   begin
+      if X > 0 then
+         declare
+            A : Integer := X;
+            B : Integer := A + 1;
+         begin
+            return B;
+         end;
+      else
+         declare
+            A : Integer := X;
+            B : Integer := A + 1;
+         begin
+            return B;
+         end;
+      end if;
+   end Dup;
+
+| ``infer.adb:4:10: medium warning: duplicate branches (Infer): code duplicated at line 11``
+
+=================
+Race Conditions
+=================
+
+-------------------------
+Race Condition Messages
+-------------------------
+
+.. container:: latex_environment
+
+   .. list-table::
+        :header-rows: 1
+
+        * - *Message*
+
+          - *Description*
+
+        * - ``unprotected access``
+
+          - Shared object access without lock
+
+        * - ``unprotected shared access``
+
+          - Object is referenced is multiple tasks
+
+        * -
+
+          - And accessed without a lock
+
+        * - ``mismatch protected access``
+
+          - Mismatch in locks used
+
+        * -
+
+          - Checked for all shared objects access
+
+        * -
+
+          - eg. task1 uses lock1, task2 uses lock2
+
+-------------------------
+Race Condition Examples
+-------------------------
+
+..
+   :toolname:`CodePeer` example (4.1.5 - race conditions)
+
+.. code:: Ada
+   :number-lines: 1
+
+   procedure Increment is
+   begin
+      Mutex_Acquire;
+      if Counter = Natural'Last then
+         Counter := Natural'First;
+      else
+         Counter := Counter + 1;
+      end if;
+      Mutex_Release;
+   end Increment;
+
+   procedure Reset is
+   begin
+      Counter := 0; -- lock missing
+   end Decrement;
+
+| ``medium warning: mismatched protected access of shared object Counter via race.increment``
+| ``medium warning: unprotected access of Counter via race.reset``
+
+=====================================
+Automatically Generated Annotations
+=====================================
+
+-----------------------
+Generated Annotations
+-----------------------
+
++ :toolname:`CodePeer` generates **annotations** on the code
++ Not errors
++ Express **properties** and **assumptions** on the code
++ Can be reviewed
+
+    + But not necessarily
+    + Can help spot **inconsistencies**
+
++ Can help understand and **debug** messages
+
+------------------------
+Annotations Categories
+------------------------
+
+.. container:: latex_environment
+
+   .. list-table::
+        :header-rows: 1
+
+        * - *Annotation*
+
+          - *Description*
+
+        * - ``precondition``
+
+          - Requirements imposed on the subprogram's inputs
+
+        * - ``presumption``
+
+          - Presumption on the result of an **external** subprogram
+
+        * - ``postcondition``
+
+          - Presumption on the outputs of a subprogram
+
+        * - ``unanalyzed call``
+
+          - External calls to unanalysed subprograms
```

Comment:
I know there's different spellings of "unanalysed" (and others) but we should be consistent :)